akka-projection-core/src/main/scala/akka/projection/internal/HandlerRecoveryImpl.scala

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.internal
+
+import scala.concurrent.Future
+import scala.util.control.NonFatal
+
+import akka.Done
+import akka.actor.ClassicActorSystemProvider
+import akka.annotation.InternalApi
+import akka.event.LoggingAdapter
+import akka.pattern.after
+import akka.pattern.retry
+import akka.projection.HandlerRecovery
+import akka.projection.HandlerRecoveryStrategy
+import akka.projection.scaladsl.SourceProvider
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object HandlerRecoveryImpl {
+
+  private val futDone = Future.successful(Done)
+
+  def applyUserRecovery[Offset, Envelope](
+      handler: HandlerRecovery[Envelope],
+      envelope: Envelope,
+      sourceProvider: SourceProvider[Offset, Envelope],
+      logger: LoggingAdapter,
+      futureCallback: () => Future[Done])(implicit systemProvider: ClassicActorSystemProvider): Future[Done] = {
+    import HandlerRecoveryStrategy.Internal._
+
+    implicit val scheduler = systemProvider.classicSystem.scheduler
+    implicit val dispatcher = systemProvider.classicSystem.dispatcher
+
+    val tryFutureCallback: () => Future[Done] = { () =>
+      try {
+        futureCallback()
+      } catch {
+        case NonFatal(e) =>
+          // in case the callback throws instead of returning failed Future
+          Future.failed(e)
+      }
+    }
+
+    // this will count as one attempt
+    val firstAttempt = tryFutureCallback()
+
+    firstAttempt.recoverWith {
+      case err =>
+        val failedOffset = sourceProvider.extractOffset(envelope)
+
+        handler.onFailure(envelope, err) match {
+          case Fail =>
+            logger.error(
+              cause = err,
+              template =
+                "Failed to process envelope with offset [{}]. Projection will stop as defined by recovery strategy.",
+              failedOffset)
+            firstAttempt
+
+          case Skip =>
+            logger.warning(
+              "Failed to process envelope with offset [{}]. " +
+              "Envelope will be skipped as defined by recovery strategy. Exception: {}",
+              failedOffset,
+              err)
+            futDone
+
+          case RetryAndFail(retries, delay) =>
+            logger.error(
```

Comment:
Should this be a warning since we're going to retry?

Reply:
yeah, I think so